### ISSUE TITLE
`unreachable` false positive when using `isinstance` on class with `__call__` when function parameter uses `ParamSpec`

### ISSUE THREAD

**Issue by DetachHead** [python/mypy#16024]
```py
from typing import Callable, ParamSpec

P = ParamSpec("P")


class Foo:
    def __call__(self) -> None:
        ...


def foo(fn: Callable[P, None]) -> None:
    if isinstance(fn, Foo):
        print("reached") # error: Statement is unreachable  [unreachable]


foo(Foo())
```
[playground](https://mypy-play.net/?mypy=latest&python=3.10&flags=show-error-context%2Cshow-error-codes%2Cstrict%2Ccheck-untyped-defs%2Cdisallow-any-decorated%2Cdisallow-any-expr%2Cdisallow-any-explicit%2Cdisallow-any-generics%2Cdisallow-any-unimported%2Cdisallow-incomplete-defs%2Cdisallow-subclassing-any%2Cdisallow-untyped-calls%2Cdisallow-untyped-decorators%2Cdisallow-untyped-defs%2Cno-implicit-reexport%2Clocal-partial-types%2Cstrict-equality%2Cwarn-incomplete-stub%2Cwarn-redundant-casts%2Cwarn-return-any%2Cwarn-unreachable%2Cwarn-unused-configs%2Cwarn-unused-ignores%2Cno-implicit-optional&gist=286f89279cc0211624590bb6e5ffbf0f)

in this example, that line is always reached at runtime

**Comment by DetachHead** [comment]
minified further:
```py
from typing import Callable

def foo(fn: Callable[[], object]) -> None:
    if isinstance(fn, type):
        print(fn)
```

**Comment by sterliakov** [comment]
Works since v1.11.0, resolved in #17389.
